sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadViaSDF.java

```diff
@@ -0,0 +1,742 @@
+    Coder<V> valueCoder = getValueCoder(coderRegistry);
+    Coder<KafkaRecord<K, V>> outputCoder = KafkaRecordCoder.of(keyCoder, valueCoder);
+    PCollection<KafkaRecord<K, V>> output =
+        input.apply(ParDo.of(new ReadFromKafkaDoFn())).setCoder(outputCoder);
+    if (isCommitOffsetEnabled() && !configuredKafkaCommit()) {
+      // TODO(BEAM-10123): Add CommitOffsetTransform to expansion.
+      LOG.warn("Offset committed is not supported yet. Ignore the value.");
+    }
+    return output;
+  }
+
+  private Coder<K> getKeyCoder(CoderRegistry coderRegistry) {
+    return (getKeyCoder() != null)
+        ? getKeyCoder()
+        : getKeyDeserializerProvider().getCoder(coderRegistry);
+  }
+
+  private Coder<V> getValueCoder(CoderRegistry coderRegistry) {
+    return (getValueCoder() != null)
+        ? getValueCoder()
+        : getValueDeserializerProvider().getCoder(coderRegistry);
+  }
+
+  private boolean configuredKafkaCommit() {
+    return getConsumerConfig().get("isolation.level") == "read_committed"
+        || Boolean.TRUE.equals(getConsumerConfig().get(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG));
+  }
+
+  static class ExtractOutputTimestampFns<K, V> {
+    public static <K, V> SerializableFunction<KafkaRecord<K, V>, Instant> useProcessingTime() {
+      return record -> Instant.now();
+    }
+
+    public static <K, V> SerializableFunction<KafkaRecord<K, V>, Instant> useCreateTime() {
+      return record -> {
+        checkArgument(
+            record.getTimestampType() == KafkaTimestampType.CREATE_TIME,
+            "Kafka record's timestamp is not 'CREATE_TIME' "
+                + "(topic: %s, partition %s, offset %s, timestamp type '%s')",
+            record.getTopic(),
+            record.getPartition(),
+            record.getOffset(),
+            record.getTimestampType());
+        return new Instant(record.getTimestamp());
+      };
+    }
+
+    public static <K, V> SerializableFunction<KafkaRecord<K, V>, Instant> useLogAppendTime() {
+      return record -> {
+        checkArgument(
+            record.getTimestampType() == KafkaTimestampType.LOG_APPEND_TIME,
+            "Kafka record's timestamp is not 'LOG_APPEND_TIME' "
+                + "(topic: %s, partition %s, offset %s, timestamp type '%s')",
+            record.getTopic(),
+            record.getPartition(),
+            record.getOffset(),
+            record.getTimestampType());
+        return new Instant(record.getTimestamp());
+      };
+    }
+  }
+
+  /**
+   * A SplittableDoFn which reads from {@link Row} in {@link KafkaSourceDescriptionSchemas} which
+   * represents a Kafka source description and outputs {@link KafkaRecord}. By default, a {@link
+   * MonotonicallyIncreasing} watermark estimator is used to track watermark.
+   */
+  @VisibleForTesting
+  class ReadFromKafkaDoFn extends DoFn<Row, KafkaRecord<K, V>> {
+
+    ReadFromKafkaDoFn() {}
+
+    private final Map<String, Object> consumerConfig = ReadViaSDF.this.getConsumerConfig();
```

Comment:
Why store these as member variables if we plan to pull them from the parent enclosed class?

Reply:
I misunderstood how nested and inner class gets serialized. I think I should mark `ReadFromKafkaDoFn` as static to avoid serializing outer class. Thanks for pointing it out!